lib/msf/ui/console/command_dispatcher/core.rb

```diff
@@ -1794,6 +1836,177 @@ def cmd_sessions(*args)
     true
   end
 
+  def get_matching_sessions(search_term)
+    matching_sessions = {}
+    terms = search_term.split
+    id_searches, type_searches, checkin_searches, searches = [], [], [], []
+
+    # Group search terms by what's being searched for
+    terms.each do |term|
+      case term.split(":").first
+      when SESSION_ID
+        id_searches << term
+      when SESSION_TYPE
+        type_searches << term
+      when LAST_CHECKIN
+        checkin_searches << term
+      else
+        print_error("Please provide valid search term. Given: #{term.split(":").first}")
+        return nil
+      end
+    end
+
+    # Group results by search term - OR filters
+    [id_searches, type_searches].each do |terms|
+      unless terms.empty?
+        id_matches = {}
+        terms.each do |term|
+          matches = filter_sessions_by_search(term)
+          return unless matches
+          id_matches = id_matches.merge(matches)
+        end
+        searches << id_matches
+      end
+    end
+
+    # Retrieve checkin search results. AND filter with a max length of 2
+    unless checkin_searches.empty?
+      unless validate_checkin_searches(checkin_searches)
+        return
+      end
+      checkin_matches = {}
+      checkin_matches = filter_sessions_by_search(checkin_searches.first)
+      if checkin_searches[1]
+        matches = filter_sessions_by_search(checkin_searches[1])
+        checkin_matches = checkin_matches.select{ |session_id, session| matches[session_id] == session }
+      end
+      searches << checkin_matches
+    end
+
+    # AND all the results together for final session list
+    unless searches.empty?
+      matching_sessions = searches.first
+      searches[1..].each do |result_set|
+        matching_sessions = matching_sessions.select{| session_id, session| result_set[session_id] == session}
+      end
+    else
+      print_error("Please provide a valid search query.")
+      return nil
+    end
+    matching_sessions
+  end
+
+  def validate_checkin_searches(checkin_searches)
+    checkin_searches.each do |search_term|
+      unless search_term.split(":").length == 3
+        print_error("Please only specify last_checkin, before or after, and a time. Ex: last_checkin:before:1m30s")
+        return false
+      end
+      time_value = search_term.split(":")[2]
+      time_unit_string = time_value.gsub(/[^a-zA-Z]/, '')
+      unless time_unit_string == time_unit_string.squeeze
```

Comment:
`squeeze` only removes duplicate letters in a row right? so `dds` becomes `ds` but it doesn't work with say for example `dmd` it would remain `dmd` right?

Reply:
Yes, but at that point it'll get caught by the parse_time function and print print_error("Unrecognized time format: #{value}")